### DISCUSSION TITLE
AIレビューの結果を技術力スコアに算入したい

### DISCUSSION THREAD

**Discussion by nunukim** [lapras-inc/lapras-user-community#21] [スコアについてのご意見・議論]
現在のLAPRASスコアv2.0のTechArticle スコアでは、Qiita のstock, ZennのLike数を算入していますが、これは記事の人気やバズを測れても、記事のクオリティを評価することはできていないので、その観点を追加するため、[AIレビュー機能](https://talent-help.lapras.com/ja/articles/8039514-ai%E3%83%AC%E3%83%93%E3%83%A5%E3%83%BC%E6%A9%9F%E8%83%BD%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)の評価結果をスコアに算入したいと考えています。

* そもそも導入すべきか
* 導入するならどのような形であるべきか

について議論したいです。

## 計算方法（案）

AIレビュー機能では、5つの観点を5点満点で評価し、それらの平均を総合点として計算している。

複数の記事の評価について、

* 「記事を書いて点数が下がる」ことがあってはならない。（単調増加性）
* 評価が低い記事をたくさん書いてもあまり点数は増えない
* 評価が特に高い記事の影響が大きくなる

上記を満たす関数の例として、

1. 評価点の最大値をとる
2. 記事ごとの評価の [LogSumExp](https://en.wikipedia.org/wiki/LogSumExp) をとる
    * これは、スケール係数 $t$ を大きく取れば、最大値を取るのと同じになる。
3. 上位N件(たとえば3件)の合計点をとる
4. $x_n$ を上位n番目の評価、 $0< r \leq 1$ をある定数として、 $\sum_{n} r^n x_n$ をとる
    * $r \to 0$ のとき、ユーザー間の大小比較という観点では 1. の評価点の最大値を取るのと同値になる。
    * $r = 1$ のとき、単純に全件の和をとったもの(3. において $N \to \infty$ としたもの)になる

1, 3, 4 については、単調減少な重み関数 $w_n \geq 0$ を使って $\sum_n w_n x_n$ という形で一般化される。
（1, 3は $n \leq N$ ならば $w_n=1$ それ以外は$w_n$ とした場合。4は $w_n = r^n$ とした場合。）

これらのどれを使うか、それとも他の関数を使うかなどの意見をいただきたい。

## 導入にあたって必要なこと

* AI レビュー記事の計算方法を決定する
* TechArticle 内におけるAIレビューの重みを決定する
* TechArticle 自体の重みを調整する

**Comment by changeworld** [comment]
そもそも

> 現在のLAPRASスコアv2.0のTechArticle スコアでは、Qiita のstock, ZennのLike数を算入していますが、これは記事の人気やバズを測れても、記事のクオリティを評価することはできていない

このように判断されているなら

> その観点を追加

観点を追加するのではなく、差し替えるべきのように思います。

「記事の人気やバズ」の要素は "技術力" よりも "影響力" の影響が非常に強く現れる(*1)ものと私は考えているため、なぜこれを "技術力" の算出に使っているのかの根拠が私にはよくわかりません。

*1: 同じ内容の記事を同じタイミングで著名人が書いたのと、無名の新人が書いたものでは、前者の方が Qiita のstock, ZennのLike数は多い(無名の新人の場合そもそも記事を閲覧してもらえる機会が少ない)

**Reply by rockymanobi** [reply]
コメントありがとうございます！今あるものが無意味なのであれば追加ではなく差し替えるべき、は本当にそうだとおもいます。
元のコメントは「AIレビューを導入しよう！」という気持ちをうけ、強く言い過ぎている感がありました :bow:
実際の認識としましては...

**「LikeやStockの数は執筆者の能力を推定するうえで一定有効と考えているものの、記事のクオリティそのものを直接評価できているわけではないのでこれを考慮にいれて、多角的な観点を取り入れてロバストなスコアに近づけたい」**

というのが正しいトーンかなと思いますので、訂正させてください :pray:

### 🗒️ LikeやStockを評価することに一定の価値はありそう、の心

【❗前提  】 **これが唯一絶対の正解である、という形で主張するものではなく、LAPRASとしてはこう考えて提供していますという、主観の入ったテキストを書きます。また、はじめからこのように考えてデザインしていたというより、「やめてみては？」と問を頂いたことで意味付けを考え、結論「まったく評価できていないわけではない」というところに落ち着いたものになります。きっかけをいただき、本当にありがとうございます。**

LAPRASとしては、直接的に測っているのは記事の人気やバズ要素ではあるものの、その数値は記事の内容や執筆者の技術力にもある程度相関するところがあると考えています。
技術力の定義とは〜というお話にもなってしまうところではありますが、多くの人に求められる知見を整理して提供できる、というのも能力の一つであり、反応ベースの考え方を能力推定に盛り込むのは妥当な面があると言えます(その他の要素も拾ってしまうだけで、まったく出鱈目な指標ではない、という考えです)。

LikeやStockが執筆者の"影響力"に大きく依存する数値であることもその通りかと思います。ただ、影響力を持つまでに至った経緯にも目を向ければ、影響力の要素が支配的であったとしても、能力とはまったく関係のない考慮から除くべきもの、という評価はできないかなと考えています。

「同じ記事を著名な人と無名の新人が書いた場合に著名人の記事が伸びる」についても、これ自体は完全に真である一方で、全体の傾向としては、無名の新人さんが著名になった方と同じ目線で記事を書くのは難しいというのも成立します。どちらの要素をtakeするかという問いになり、今のところは後者を取るべきかなと考えている次第です。

もちろん、インプレッションを伸ばすために意図的に誤りやツッコミどころを含める等のハック可能な余地があったり、インフルエンサー的な活動に極フリの方のスコアが高く出てしまう部分もあります。こちらは影響が大きくなれば対処を考えていく必要があると認識しています。

よって、Likeやストックを評価する軸を捨てるのではなく維持し、評価に組み込めていなかった「記事のクオリティ」を計算に盛り込むことで、多角的に評価ができる状態に近づけたい、というのがAIレビュー算入の方向性として今考えていることとなります。

※TechArticle全体やLike,Stock要素の重みの調整などは納得感を得られるようコミュニティ進めていきたいと思っており、広く意見いただきたいとも思っております。

###  🗒️ 別のお話

話が変わりますが、自らの発信力だけで記事のインプレッションを稼げるようにまでなった人は外れ値的な存在なのかもしれません。なので、個人の影響力によってどんな記事でも拡散させることができる人の"影響力差し引いた技術力"を正しく評価する指標としたいのか、「コミュニティ/第三者に評価されるステップを踏むことではじめて記事が拡散されるような、大多数の人に適用したときにうまく動く形にしたいのか...など、そういったたぐいの話も考えないといけないなと問いを頂いて感じたところです。ありがとうございます 🙏

**Reply by changeworld** [reply]
@rockymanobi 
ご回答ありがとうございました。

LAPRASとして技術力として記事の Like や Stock を計上するということ結論について理解いたしました。

**Comment by nunukim** [comment]
AIレビューのスコア算入について、社内での検討結果を共有します。
以下の仕様で次回のマイナーバージョンアップ(v2.2.0)への追加を予定しています。

# AIレビューのスコア算入仕様

* Raw AIレビュースコアは以下の式で定義する:
     $$\displaystyle \mathrm{(RawAiReviewScore)} = \frac{1}{t} \mathrm{LSE}_n \left( t \max(s_n - \theta, 0) \right) \\;\\; \cdots (1) $$
    * $\mathrm{LSE}\_n(x\_n) = \log \sum_{n=1}^N \exp (x_n)$ は [LogSumExp 関数](https://en.wikipedia.org/wiki/LogSumExp)
    * logは自然対数
    * N はAIレビューで評価された記事数
    * $s_n$ は各記事のAIレビューの点数 (総合点)
    * 点数の閾値 θ = 3
    * パラメータ t = 5

* Raw AI レビュースコアとこれまでのRaw Tech Articleスコア (仮称: Popularityスコア) から、以下の式で新しいTechArticleスコアを定義する
    ```
    RawTechArticleScore = RawTechArticlePopularityScore + W_AI_REVIEW * RawAiReviewScore
    ```
    * 重み `W_AI_REVIEW` = 10

<img width="785" alt="スクリーンショット 2025-02-20 21 07 42" src="https://github.com/user-attachments/assets/ccee297e-2865-4e7a-853a-65d8e85fa3fa" />


# 考え方

## 評価関数の形式

AIレビューの点数を集計する際の、[元投稿](https://github.com/lapras-inc/lapras-user-community/discussions/21#discussion-7711422)における1~4の方式は、TechArticle やGitHub における複数の記事/リポジトリの評価値を集計する際の集計関数の議論( #24 )と同様となる。（1,3,4 が順位ベースの考え方、2が値ベースの考え方となる。）

今回は、以下の理由から値ベースの LogSumExp を使う方式を採用する。（上述の式（1））

* 「質の高い記事を書き続けている」ことを重視するのであれば、上位の記事だけでなく、なるべく多くの記事を参照すべき
* 値ベースの方が順位ベースよりロジックが単純で理解しやすく、値の変化も予測しやすい。
* LogSumExp は「点数の最大値」と「記事の件数（の対数）」を程よく反映したものになっている（後述）


## 点数の閾値θ

AIレビューの点数の値はほぼ全てが3.0~4.2の範囲に収まっている。

![スクリーンショット 2025-02-17 22 36 44](https://github.com/user-attachments/assets/8ff49ae6-64bf-4bd3-9cd1-dfc43bea2436)

このため、「どんなに質の低い記事でも最低3.0点はもらえる」状態となり、この値をそのままRawスコアとして用いてしまうと「記事を全く書いていない人」と「テスト投稿のような内容のない記事を1本だけ書いている人」との間の点数が開き過ぎてしまう。
この問題を回避するため、Rawスコアに用いる値はAIレビューの評価点からθ=3.0を減じたものとし、3.0点を下回るものについてはスコアの対象としないこととした。

## スケーリングパラメータt

(1)式において、スケーリングパラメータ tを用いることで、「AIレビューの最高点」を重視するか「記事数（の対数）」を重視するかのトレードオフを調整することができる。

実際、

* $t \to \infty$ のとき、 $\frac{1}{t}\mathrm{LSE}\_n(t x\_{n}) \to \max\_n (x\_n)$ より、(1)式は「AIレビューの最高点」の値に近づく。
* $t \to 0$ のとき、 $\frac{1}{t}\mathrm{LSE}\_n(t x\_{n}) \approx \frac{\log(N)}{t}$ より、(2)式は記事数の対数に比例した値に近づく。

「質の高い記事を出し続ける」ことを評価する場合、最高点と記事数の両方を程よく評価するようにtを設定する必要がある。
たとえば、tが大きすぎると、最高点を取った記事のみが点数に影響するため、「たまたま1つだけ点数が高い記事がある人」と「点数が高い記事をたくさん書いている人」を同じように評価してしまうという問題がある。また、最高点を取った1つの記事のみでスコアが決まってしまうため、AIレビューの判定誤差に対して不安定になってしまうという問題がある。
一方で、tが小さすぎると、もはや質は関係なく、記事数だけでスコアが決まることになってしまう。

実際にtの値を変化させたとき「Raw AI レビュースコア」に対する「記事数の対数」「レビュー最高点」の相関係数がそれぞれどのように変化するかを下図に示す。

<img width="449" alt="スクリーンショット 2025-02-18 0 29 38" src="https://github.com/user-attachments/assets/b37f88ff-2c15-485e-abff-0357fe5a5d7b" />


今回は、「最高点と記事数の両方を程よく評価する」という観点から、 t=5 という値を採用する。
これは、「評価4.0点の記事1本」が「3.8点の記事2.7(= $e$ )本」や「3.0点の記事148.4(= $e^5$ )本」と等価であるとすることに相当する。

## 重み `W_AI_REVIEW`

最終的なスコアへの影響を考慮し、既存のRawTechArticlePopularity スコアとの相対的な重みの値を `W_AI_REVIEW` = 10 とした。

このとき、AIレビューのスコアへの寄与がPopularityの寄与に比べてどの程度の大きさかを示す。

たとえば、記事数が3件で、これらのStock/Like数とAIレビュー点数がそれぞれ全て等しいとしたとき、Popularity とAIレビューの寄与が等しくなる条件は以下のようになる。

![スクリーンショット 2025-02-21 19 59 02](https://github.com/user-attachments/assets/786329b7-3f04-4ee0-b56b-cc79bcb5b33b)

* AIレビュー3.0点 ≒ Stock/Like 2件
* AIレビュー4.0点 ≒ Stock/Like 5件

また、AIレビュー4.0点の記事が多数存在するとした時の条件は以下のようになる。


![スクリーンショット 2025-02-21 20 06 59](https://github.com/user-attachments/assets/4e752d4b-7863-4cf9-91fc-632fd885834b)

* AIレビュー4.0点の記事10本 ≒ Stock/Like 10.5件
* AIレビュー4.0点の記事40本 ≒ Stock/Like 12.3件

これらのことから、AIレビューによるスコアへの寄与はおおむね「記事が全く閲覧されなかったとしても、質が高い記事は stock/like 2~5回（記事数が多い場合は12回）程度がついているものとして評価する」ことに相当するものといえる。


# スコア変動

今回のロジック修正によるスコア変動は以下の通り。

図: AIレビューで評価された記事数ごとの各スコア帯におけるスコアの変動

![スクリーンショット 2025-02-21 20 22 48](https://github.com/user-attachments/assets/b4a3c7c6-0e32-482b-88ba-7104b573148c)



* AIレビューによる記事評価が存在する人は評価に応じてスコアが上昇する
    * とくに、もともとTechArticleスコアが低い人ほど、新たにAIレビューが評価されることの影響が大きい。
    * 記事数が多い人は、もともとのTechArticleスコアも高い傾向がありAIレビューの評価の影響が相対的に大きくないため、記事数の多さはスコアの上昇幅にあまり影響しない。
* AIレビューの記事評価が存在しない人は、相対的にスコアが減少する。
    * 全体的に 0.01〜0.03程度減少する。
    * スコア3.6程度の人が最も減少幅が大きく、0.03程度低下する。

**Comment by ryo-endo** [comment]
AIレビューの結果を技術力スコアに算入する変更は、上記の内容で来週3月13日ごろに実際のスコアに反映される予定です。
